actors/builtin/market/testing.go

```diff
@@ -0,0 +1,229 @@
+		"negative total client locked collateral: %v", st.TotalClientLockedCollateral)
+
+	acc.Require(
+		st.TotalProviderLockedCollateral.GreaterThanEqual(big.Zero()),
+		"negative total provider locked collateral: %v", st.TotalClientLockedCollateral)
+
+	acc.Require(
+		st.TotalClientStorageFee.GreaterThanEqual(big.Zero()),
+		"negative total client storage fee: %v", st.TotalClientLockedCollateral)
+
+	//
+	// Proposals
+	//
+
+	allIDs := make(map[abi.DealID]struct{})
+	proposalCids := make(map[cid.Cid]struct{})
+	maxDealID := int64(-1)
+	proposalCount := uint64(0)
+
+	proposals, err := adt.AsArray(store, st.Proposals)
+	if err != nil {
+		return nil, acc, err
+	}
+	var proposal DealProposal
+	err = proposals.ForEach(&proposal, func(dealID int64) error {
+		allIDs[abi.DealID(dealID)] = struct{}{}
+
+		pcid, err := proposal.Cid()
+		if err != nil {
+			return err
+		}
+
+		// keep some state
+		proposalCids[pcid] = struct{}{}
+		if dealID > maxDealID {
+			maxDealID = dealID
+		}
+		proposalCount++
+
+		return nil
+	})
+	if err != nil {
+		return nil, acc, err
+	}
+
+	// next id should be higher than any existing deal
+	acc.Require(int64(st.NextID) > maxDealID, "next id, %d, is not greater than highest id in proposals, %d", st.NextID, maxDealID)
+
+	//
+	// Deal States
+	//
+
+	dealStateCount := uint64(0)
+	dealStates, err := adt.AsArray(store, st.States)
+	if err != nil {
+		return nil, acc, err
+	}
+	var dealState DealState
+	err = dealStates.ForEach(&dealState, func(dealID int64) error {
+		acc.Require(
+			dealState.SectorStartEpoch >= 0,
+			"deal state start epoch undefined %d: %v", dealID, dealState)
+
+		acc.Require(
+			dealState.LastUpdatedEpoch == epochUndefined || dealState.LastUpdatedEpoch >= dealState.SectorStartEpoch,
+			"deal state last updated before sector start %d: %v", dealID, dealState)
+
+		acc.Require(
+			dealState.SlashEpoch == epochUndefined || dealState.SlashEpoch >= dealState.SectorStartEpoch,
+			"deal state slashed before sector start %d: %v", dealID, dealState)
+
+		_, found := allIDs[abi.DealID(dealID)]
+		acc.Require(found, "deal state references deal %d not found in proposals", dealID)
+
+		dealStateCount++
+		return nil
+	})
+	if err != nil {
+		return nil, acc, err
+	}
+
+	//
+	// Pending Proposals
+	//
+
+	pendingProposalCount := uint64(0)
+	pendingProposals, err := adt.AsMap(store, st.PendingProposals)
+	if err != nil {
+		return nil, nil, err
+	}
+	var pendingProposal DealProposal
+	err = pendingProposals.ForEach(&pendingProposal, func(key string) error {
+		proposalCID, err := cid.Parse([]byte(key))
+		if err != nil {
+			return err
+		}
+
+		pcid, err := pendingProposal.Cid()
+		if err != nil {
+			return err
+		}
+		acc.Require(pcid.Equals(proposalCID), "pending proposal's key does not match its CID %v != %v", pcid, proposalCID)
+
+		_, found := proposalCids[pcid]
+		acc.Require(found, "pending proposal with cid %v not fond within proposals %v", pcid, pendingProposals)
+
+		pendingProposalCount++
```

Comment:
Can we check it's being garbage collected? I think there should be no pending deal with start date < `current - DealUpdatesInterval`

Reply:
This one is kind of tough. Garbage collection happens in cron. So there technically can be a period between expiration and the next cron run where it doesn't hold, unless you assume crons are run for null blocks (which I think we are).

More to the point, we don't require our unit tests to run cron for every epoch, so adding this requirement causes many tests to fail, and it would be a fair amount of work and potentially a huge performance hit to start imposing that requirement.